### ISSUE TITLE
Insert and delete? And some other questions

### ISSUE THREAD

**Issue by schneems** [greensync/interval-tree#16]
Hello, I think am in need of an interval tree. I saw yours but it looks like it only handles a fixed set of ranges rather than allowing for inserts and deletes. Is that correct?

## Context

Ultimately I am working on this https://twitter.com/schneems/status/1479554521307176960 where I want to speed up this code: https://github.com/zombocom/dead_end/blob/0170bf3df9c045e4d5f042c6b161ff0c30f2efda/lib/dead_end/code_frontier.rb#L152-L154. I need to find all elements in a collection that are between two values.

I've begun attempting to write an augmented interval tree based on a red-black tree and then specializing it to optimize for querying elements that are completely within a given range. I'm doing one insert and one search for every operation so both need to be equally fast. Right now with a plain binary tree my insert is fast, but searching for contained entries is slower than a naive array iteration of N elements. 

It's my belief that adding annotations to the red-black tree, (and possibly reversing the sort interval, so it's sorted on end key first and start key last) will allow me to eliminate more nodes from my search and hopefully give me better than naive time. 

## Questions

- Can you verify that insert/delete are not implemented on this?
- Assuming it could insert/delete would this centered interval tree be a good structure for querying elements contained by a range?
- Any general advice, hints, or tips? Sorry for the random extra questions, I appreciate that this isn't stack overflow but wanted to see if you had any thoughts

BTW thanks for open-sourcing this.

**Comment by amarzot** [comment]
@schneems 

* Insert/delete are not implemented here. Don't think there are plans to implement them (this could change, but current maintainers have been MIA for a while).
* Querying by range is one of the modes supported by this implementation
*  Seeing as you need to often insert and delete items from the tree, this might not be the implementation for you as it currently exists.

I poked around your code base for a bit, but it looks more complicated than I have the time to dive into right now. I think I have an okay idea of what you're trying to do, but not sure if I understand enough to give an informed opinion, sorry. I also realize the response is many months late, but figured I'd send one anyways.

**Comment by schneems** [comment]
Thanks for the response. Regarding this perf attempt, this is still a hotspot in the code, however I don’t think that this is a good optimization candidate.

I’ve got a long running branch where instead I’m working on a whole new implementation of the project. The core difference is that the existing search for syntax errors searches and builds a tree(ish) structure but it’s destructive and irreversible. The new implementation divides building a tree (graph really) and searching it. I designed it so that every step in building a node captures prior nodes so that way we don’t have to query for overlap (as each step guarantees overlap and we know exactly which nodes, we don’t have to hunt them down in a separate data structure.

The new code is a long way off from being done though. If you get inspired to look at what I linked to and have some ideas you think might be viable then im happy to look again. Otherwise it’s not urgent.
